modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -143,30 +148,45 @@ public void stop() throws Exception {
 
     /** {@inheritDoc} */
     @Override
-    public CompletableFuture<QueryPlan> prepareAsync(SqlNode sqlNode, BaseQueryContext ctx) {
-        try {
-            assert single(sqlNode);
+    public CompletableFuture<QueryPlan> prepareAsync(SqlNode sqlNode, BaseQueryContext ctx, long plannerTimeout) {
+        CompletableFuture<QueryPlan> result;
+
+        assert single(sqlNode);
+
+        SqlQueryType queryType = Commons.getQueryType(sqlNode);
+        assert queryType != null : "No query type for query: " + sqlNode;
+
+        PlanningContext planningContext = PlanningContext.builder()
+                .parentContext(ctx)
+                .query(ctx.query())
+                .plannerTimeout(plannerTimeout)
+                .build();
 
-            SqlQueryType queryType = Commons.getQueryType(sqlNode);
-            assert queryType != null : "No query type for query: " + sqlNode;
+        result = prepareAsync0(sqlNode, queryType, planningContext);
 
-            PlanningContext planningContext = PlanningContext.builder()
-                    .parentContext(ctx)
-                    .build();
+        return result.exceptionally(ex -> {
+            Throwable th = ExceptionUtils.unwrapCause(ex);
+            if (planningContext.timeouted() && th instanceof RelOptPlanner.CannotPlanException) {
+                LOG.info("Query plan is absent due to planing timeout is reached [query={}]", ctx.query());
```

Comment:
so, we are going to have 2 lines in the log per each timeout?

Reply:
Why is exception not enough?

Reply:
seems you are right. Will be removed